File: src/main/agent/AgentBridge.ts
/**
 * AgentBridge - WebSocket 服务端，运行在 Electron 主进程
 * MCP Server (stdio 子进程) 通过 WS 连接此桥接器，转发请求到 AgentManager
 * @author weibin
 */

import { EventEmitter } from 'events'
import { WebSocketServer, WebSocket } from 'ws'
import type { BridgeRequest, BridgeResponse } from './types'

export class AgentBridge extends EventEmitter {
  private wss: WebSocketServer | null = null
  private connections: Map<string, WebSocket> = new Map() // sessionId → ws
  private port: number = 0
  private teamBridgeHandler?: (request: BridgeRequest) => Promise<{ result?: any; error?: string }>
  
  // ★ 心跳检测配置
  private heartbeatInterval: NodeJS.Timeout | null = null
  private readonly HEARTBEAT_INTERVAL_MS = 30000 // 30 秒
  private readonly HEARTBEAT_TIMEOUT_MS = 10000 // 10 秒超时

  /**
   * 启动 WebSocket 服务
   */
  start(port: number): void {
    this.port = port
    this.wss = new WebSocketServer({ host: '127.0.0.1', port })

    this.wss.on('connection', (ws: WebSocket) => {
      let registeredSessionId: string | null = null

      ws.on('message', (raw: Buffer | string) => {
        try {
          const msg = JSON.parse(raw.toString())

          // 注册消息：MCP Server 连接后先注册 sessionId
          if (msg.type === 'register') {
            registeredSessionId = msg.sessionId
            this.connections.set(msg.sessionId, ws)
            console.log(`[AgentBridge] MCP Server registered for session: ${msg.sessionId}`)
            ws.send(JSON.stringify({ type: 'registered', sessionId: msg.sessionId }))
            return
          }

          // 文件变更事件：MCP Server 本地执行文件操作后通知
          if (msg.type === 'file-change') {
            this.emit('file-change', {
              sessionId: registeredSessionId || msg.sessionId,
              data: msg.data,
            })
            return
          }

          // 请求消息：转发到 AgentManager 处理（team_* 方法由 TeamBridge 处理）
          if (msg.type === 'request') {
            const request: BridgeRequest = {
              id: msg.id,
              sessionId: registeredSessionId || msg.sessionId,
              method: msg.method,
              params: msg.params || {}
            }

            // ★ team_* 方法由 TeamBridge 处理
            if (
              this.teamBridgeHandler &&
              (request.method === 'team_message_role' ||
                request.method === 'team_broadcast' ||
                request.method === 'team_claim_task' ||
                request.method === 'team_complete_task' ||
                request.method === 'team_get_tasks' ||
                request.method === 'team_get_members')
            ) {
              this.teamBridgeHandler(request).then(result => {
                if (ws.readyState === WebSocket.OPEN) {
                  ws.send(JSON.stringify({ type: 'response', id: request.id, result: result.result, error: result.error }))
                }
              }).catch(err => {
                if (ws.readyState === WebSocket.OPEN) {
                  ws.send(JSON.stringify({ type: 'response', id: request.id, error: err.message }))
                }
              })
              return
            }

            this.emit('request', request, (response: BridgeResponse) => {
              if (ws.readyState === WebSocket.OPEN) {
                ws.send(JSON.stringify({ type: 'response', ...response }))
              }
            })
          }
        } catch (err) {
          console.error('[AgentBridge] Failed to parse message:', err)
        }
      })

      ws.on('close', () => {
        if (registeredSessionId) {
          this.connections.delete(registeredSessionId)
          console.log(`[AgentBridge] MCP Server disconnected: ${registeredSessionId}`)
        }
      })

      ws.on('error', (err) => {
        console.error('[AgentBridge] WebSocket error:', err)
      })
      
      // ★ 初始化心跳标记
      ;(ws as any).isAlive = true
      ws.on('pong', () => {
        ;(ws as any).isAlive = true
      })
    })

    this.wss.on('error', (err) => {
      console.error(`[AgentBridge] Server error on port ${port}:`, err)
    })

    // ★ 启动心跳检测定时器
    this.startHeartbeat()

    console.log(`[AgentBridge] WebSocket server started on 127.0.0.1:${port}`)
  }

  /**
   * 注册团队通信处理器（TeamManager 初始化后调用）
   */
  setTeamBridgeHandler(
    handler: (request: BridgeRequest) => Promise<{ result?: any; error?: string }>
  ): void {
    this.teamBridgeHandler = handler
  }

  /**
   * ★ 启动心跳检测
   */
  private startHeartbeat(): void {
    if (this.heartbeatInterval) {
      clearInterval(this.heartbeatInterval)
    }

    this.heartbeatInterval = setInterval(() => {
      if (!this.wss) return

      this.wss.clients.forEach((ws: WebSocket) => {
        // 如果客户端上次没有响应 ping，则断开连接
        if ((ws as any).isAlive === false) {
          console.warn('[AgentBridge] Client failed heartbeat, terminating connection')
          return ws.terminate()
        }

        // 标记为未响应，发送 ping
        ;(ws as any).isAlive = false
        ws.ping()
      })
    }, this.HEARTBEAT_INTERVAL_MS)
  }

  /**
   * ★ 停止心跳检测
   */
  private stopHeartbeat(): void {
    if (this.heartbeatInterval) {
      clearInterval(this.heartbeatInterval)
      this.heartbeatInterval = null
    }
  }

  /**
   * 获取当前端口
   */
  getPort(): number {
    return this.port
  }

  /**
   * 关闭服务
   */
  close(): void {
    // ★ 停止心跳检测
    this.stopHeartbeat()
    
    if (this.wss) {
      // 关闭所有连接
      for (const ws of this.connections.values()) {
        try { ws.close() } catch (_) { /* ignore */ }
      }
      this.connections.clear()
      this.wss.close()
      this.wss = null
      console.log('[AgentBridge] Server closed')
    }
  }
}

File: src-tauri/src/services/agent_bridge.rs
//! AgentBridge WebSocket server service
//!
//! Phase 3: Replace Node.js AgentBridge.ts with Rust implementation.
//! Provides MCP (Model Context Protocol) communication over WebSocket.
//!
//! Features:
//! - WebSocket server on port 63721
//! - JSON-RPC 2.0 protocol
//! - Session registration and management
//! - Authentication with Bearer token
//! - Request routing to AgentManager
//! - Heartbeat detection

use futures_util::{SinkExt, StreamExt};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::net::TcpListener;
use tokio::sync::{mpsc, RwLock};
use tokio_tungstenite::tungstenite::Message;
use tracing::{debug, error, info, warn};
use uuid::Uuid;

/// Bridge request from MCP client
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BridgeRequest {
    pub id: String,
    pub session_id: String,
    pub method: String,
    #[serde(default)]
    pub params: serde_json::Value,
}

/// Bridge response to MCP client
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct BridgeResponse {
    pub id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Registered MCP client connection
struct McpConnection {
    /// WebSocket sender
    tx: mpsc::UnboundedSender<Message>,
    /// Session ID
    session_id: String,
    /// Last heartbeat timestamp
    last_heartbeat: std::time::Instant,
}

pub struct AgentBridgeService {
    /// Registered connections: session_id -> connection
    connections: Arc<RwLock<HashMap<String, McpConnection>>>,
    /// Server port
    port: u16,
    /// Auth token for MCP clients
    auth_token: String,
    /// Request handler callback
    request_handler: Option<Arc<dyn Fn(BridgeRequest) -> Result<serde_json::Value, String> + Send + Sync>>,
}

impl AgentBridgeService {
    pub fn new() -> Self {
        // Generate random auth token (32 bytes, base64url encoded)
        let auth_token = Uuid::new_v4().to_string() + &Uuid::new_v4().to_string();
        info!("AgentBridge auth token generated (first 8 chars: {}...)", &auth_token[..8]);

        Self {
            connections: Arc::new(RwLock::new(HashMap::new())),
            port: 63721,
            auth_token,
            request_handler: None,
        }
    }

    /// Set request handler callback
    pub fn set_request_handler<F>(&mut self, handler: F)
    where
        F: Fn(BridgeRequest) -> Result<serde_json::Value, String> + Send + Sync + 'static,
    {
        self.request_handler = Some(Arc::new(handler));
    }

    /// Start the WebSocket server
    pub async fn start(&self, port: u16) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let addr = format!("127.0.0.1:{}", port);
        let listener = TcpListener::bind(&addr).await?;
        info!("AgentBridge WebSocket server listening on {}", addr);

        let connections = self.connections.clone();
        let auth_token = self.auth_token.clone();
        let request_handler = self.request_handler.clone();

        // Spawn heartbeat checker
        let conn_clone = connections.clone();
        tokio::spawn(async move {
            Self::heartbeat_loop(conn_clone).await;
        });

        // Accept connections
        loop {
            match listener.accept().await {
                Ok((stream, addr)) => {
                    info!("New connection from {}", addr);
                    
                    let connections = connections.clone();
                    let auth_token = auth_token.clone();
                    let request_handler = request_handler.clone();

                    tokio::spawn(async move {
                        if let Err(e) = Self::handle_connection(
                            stream,
                            connections,
                            auth_token,
                            request_handler,
                        ).await {
                            error!("Connection error: {}", e);
                        }
                    });
                }
                Err(e) => {
                    error!("Failed to accept connection: {}", e);
                }
            }
        }
    }

    /// Handle a single WebSocket connection
    async fn handle_connection(
        stream: tokio::net::TcpStream,
        connections: Arc<RwLock<HashMap<String, McpConnection>>>,
        auth_token: String,
        request_handler: Option<Arc<dyn Fn(BridgeRequest) -> Result<serde_json::Value, String> + Send + Sync>>,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        use tokio_tungstenite::accept_async;
        
        let ws_stream = accept_async(stream).await?;
        let (mut write, mut read) = ws_stream.split();

        // Create channel for sending messages
        let (tx, mut rx) = mpsc::unbounded_channel::<Message>();

        // Spawn writer task
        let writer_task = tokio::spawn(async move {
            while let Some(msg) = rx.recv().await {
                if write.send(msg).await.is_err() {
                    break;
                }
            }
        });

        // Read messages
        let mut registered_session_id: Option<String> = None;

        while let Some(msg_result) = read.next().await {
            match msg_result {
                Ok(msg) => {
                    if let Message::Text(text) = msg {
                        if let Err(e) = Self::handle_message(
                            &text,
                            &connections,
                            &auth_token,
                            &tx,
                            &mut registered_session_id,
                            &request_handler,
                        ).await {
                            warn!("Message handling error: {}", e);
                        }
                    } else if let Message::Close(_) = msg {
                        info!("Client closed connection");
                        break;
                    }
                }
                Err(e) => {
                    error!("WebSocket error: {}", e);
                    break;
                }
            }
        }

        // Clean up connection
        if let Some(session_id) = registered_session_id {
            connections.write().await.remove(&session_id);
            info!("Removed connection for session {}", session_id);
        }

        writer_task.abort();
        Ok(())
    }

    /// Handle a single message
    async fn handle_message(
        text: &str,
        connections: &Arc<RwLock<HashMap<String, McpConnection>>>,
        auth_token: &str,
        tx: &mpsc::UnboundedSender<Message>,
        registered_session_id: &mut Option<String>,
        request_handler: &Option<Arc<dyn Fn(BridgeRequest) -> Result<serde_json::Value, String> + Send + Sync>>,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        // Parse JSON
        let value: serde_json::Value = serde_json::from_str(text)?;

        // Handle register message
        if let Some(type_field) = value.get("type").and_then(|v| v.as_str()) {
            match type_field {
                "register" => {
                    if let Some(session_id) = value.get("sessionId").and_then(|v| v.as_str()) {
                        // Check authentication (should be done at HTTP upgrade level)
                        // For now, just register
                        *registered_session_id = Some(session_id.to_string());

                        let conn = McpConnection {
                            tx: tx.clone(),
                            session_id: session_id.to_string(),
                            last_heartbeat: std::time::Instant::now(),
                        };

                        connections.write().await.insert(session_id.to_string(), conn);
                        info!("MCP Server registered for session: {}", session_id);

                        // Send registered response
                        let response = serde_json::json!({
                            "type": "registered",
                            "sessionId": session_id
                        });
                        tx.send(Message::Text(response.to_string().into()))?;
                    }
                }
                "file-change" => {
                    // Forward file-change event to main process
                    if let Some(session_id) = registered_session_id {
                        info!("File change event from session {}", session_id);
                        // TODO: Emit event to main process
                    }
                }
                "request" => {
                    // Handle JSON-RPC request
                    if let Some(request_handler) = request_handler {
                        let bridge_request = BridgeRequest {
                            id: value.get("id").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                            session_id: registered_session_id.clone().unwrap_or_default(),
                            method: value.get("method").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                            params: value.get("params").cloned().unwrap_or(serde_json::Value::Null),
                        };

                        // Process request
                        let result = request_handler(bridge_request.clone());

                        // Send response
                        let response = match result {
                            Ok(result_value) => BridgeResponse {
                                id: bridge_request.id,
                                result: Some(result_value),
                                error: None,
                            },
                            Err(error) => BridgeResponse {
                                id: bridge_request.id,
                                result: None,
                                error: Some(error),
                            },
                        };

                        let response_json = serde_json::to_string(&response)?;
                        tx.send(Message::Text(response_json.into()))?;
                    }
                }
                _ => {
                    debug!("Unknown message type: {}", type_field);
                }
            }
        }

        Ok(())
    }

    /// Heartbeat checker loop
    async fn heartbeat_loop(connections: Arc<RwLock<HashMap<String, McpConnection>>>) {
        let mut interval = tokio::time::interval(tokio::time::Duration::from_secs(30));

        loop {
            interval.tick().await;

            let now = std::time::Instant::now();
            let mut to_remove = Vec::new();

            {
                let conns = connections.read().await;
                for (session_id, conn) in conns.iter() {
                    if now.duration_since(conn.last_heartbeat).as_secs() > 60 {
                        to_remove.push(session_id.clone());
                    }
                }
            }

            if !to_remove.is_empty() {
                let mut conns = connections.write().await;
                for session_id in to_remove {
                    conns.remove(&session_id);
                    warn!("Removed stale connection for session {}", session_id);
                }
            }
        }
    }

    /// Get auth token (for MCP client configuration)
    pub fn get_auth_token(&self) -> &str {
        &self.auth_token
    }

    /// Get server port
    pub fn get_port(&self) -> u16 {
        self.port
    }

    /// Get number of active connections
    pub async fn get_connection_count(&self) -> usize {
        self.connections.read().await.len()
    }
}
